File: tests/typesafety/python/function_values_misuse.py
# expect: fails
# title: A key that returns the item name, rejected by the checker
# misuse-of: function_values
from collections.abc import Callable
from dataclasses import dataclass
from typing import NewType

Pence = NewType("Pence", int)


@dataclass(frozen=True)
class InvoiceLine:
    item: str
    pence: Pence
    quantity: int


def picking_list(
    lines: list[InvoiceLine], key: Callable[[InvoiceLine], int]
) -> list[InvoiceLine]:
    return sorted(lines, key=key)


by_name = picking_list([], lambda line: line.item)

File: tests/typesafety/typescript/function_values_misuse.ts
// expect: fails

declare const penceBrand: unique symbol;
type Pence = number & { readonly [penceBrand]: true };

type InvoiceLine = { readonly item: string; readonly pence: Pence; readonly quantity: number };

function pickingList(
  lines: readonly InvoiceLine[],
  key: (line: InvoiceLine) => number,
): InvoiceLine[] {
  return [...lines].sort((a, b) => key(a) - key(b));
}

export const byName = pickingList([], (line) => line.item);
